model.go

```diff
@@ -0,0 +1,265 @@
+	return m.Machines[m.MachineMap[id]]
+}
+
+func (m *Model) getUnitMachine(appName string, index int) string {
+	if m == nil || m.Applications == nil {
+		return ""
+	}
+	app := m.Applications[appName]
+	if app == nil {
+		return ""
+	}
+	target := fmt.Sprintf("%s/%d", appName, index)
+	for _, unit := range app.Units {
+		if unit.Name == target {
+			return unit.Machine
+		}
+	}
+	return ""
+}
+
+// Application represents an existing charm deployed in the model.
+type Application struct {
+	Name        string
+	Charm       string // The charm URL.
+	Options     map[string]interface{}
+	Annotations map[string]string
+	Constraints string // TODO: not updated yet.
+	Exposed     bool
+	// TODO: handle changes in:
+	//   endpoint bindings -- possible even?
+	//   storage
+	//   series
+
+	Units []Unit
+}
+
+type Unit struct {
+	Name    string
+	Machine string
+}
+
+// Machine represents an existing machine in the model.
+type Machine struct {
+	ID          string
+	Annotations map[string]string
+}
+
+func (m *Model) hasCharm(charm string) bool {
+	if m == nil || len(m.Applications) == 0 {
+		return false
+	}
+	for _, app := range m.Applications {
+		if app.Charm == charm {
+			return true
+		}
+	}
+	return false
+}
+
+func (m *Model) getApplication(name string) *Application {
+	if m == nil {
+		return nil
+	}
+	return m.Applications[name]
+}
+
+func (m *Model) unitMachinesWithoutApp(sourceApp, targetApp, container string) []string {
+	source := m.getApplication(sourceApp)
+	if source == nil {
+		return []string{}
+	}
+
+	target := m.getApplication(targetApp)
+	machines := set.NewStrings()
+	for _, unit := range source.Units {
+		machines.Add(topLevelMachine(unit.Machine))
+	}
+	if target != nil {
+		for _, unit := range target.Units {
+			if container == "" {
+				machines.Remove(unit.Machine)
+			} else {
+				machineTag := names.NewMachineTag(unit.Machine)
+				if machineTag.ContainerType() == container {
+					machines.Remove(machineTag.Parent().Id())
+				}
+			}
+		}
+	}
+
+	return utils.SortStringsNaturally(machines.Values())
+}
+
+func (a *Application) unitCount() int {
+	if a == nil {
+		return 0
+	}
+	return len(a.Units)
+}
+
+func (a *Application) changedAnnotations(annotations map[string]string) map[string]string {
+	if a == nil || len(a.Annotations) == 0 {
+		return annotations
+	}
+	changes := make(map[string]string)
+	for key, value := range annotations {
+		current, found := a.Annotations[key]
+		if !found || current != value {
+			changes[key] = value
+		}
+	}
+	return changes
+}
+
+func (a *Application) changedOptions(options map[string]interface{}) map[string]interface{} {
+	if a == nil || len(a.Options) == 0 {
+		return options
+	}
+	changes := make(map[string]interface{})
+	for key, value := range options {
+		current, found := a.Options[key]
+		// options should have been validated by now to only contain comparable
+		// types.
+		if !found || current != value {
+			changes[key] = value
```

Comment:
Float vs int comparison for when one has come from JSON and the other from YAML.

Also this will need to keep old values so you can report the old one with the new one.

Reply:
done, but not keeping old values